### DISCUSSION TITLE
Unexpected base64 output ... worth filing an Issue?

### DISCUSSION THREAD

**Discussion by jpluscplusm** [cue-lang/cue#1639] [Q&A] [answered]
I was trying to figure out how to quote a potentially hyphenated component of a nested key, at the shell, when this happened:

```
$ cat foo.cue 
foo: {
        wibble:    123
        "bar-baz": 11
}
$ X=wibble
$ cue export foo.cue -e "foo.\"$X\"" 
123
$ X=bar-baz
$ cue export foo.cue -e "foo.\"$X\"" 
11
$ cue export foo.cue  -e 'foo."$X"'
undefined field: $X:
    --expression:1:5
############ So far, so good. I understand everything above this line ##############
$ cue export foo.cue  -e \'foo."$X"\'
"Zm9vLmJhci1iYXo="
$ cue export foo.cue  -e \'foo."$X"\' --out text
cannot use value 'foo.bar-baz' (type bytes) as string:
    --expression:1:1
$ cue export foo.cue  -e \'foo."$X"\' | tr -d "'" 
"Zm9vLmJhci1iYXo="
$ cue export foo.cue  -e \'foo."$X"\' | tr -d '"' | base64 -d ; echo 
foo.bar-baz
```

Now, I *know* that I got the quoting wrong. I know I actually needed the `"foo.\"$X\""` form, from near the top of the session.

I also /suspect/ that the very unexpected base64 output is explainable ... but that doesn't stop it being really surprising!

Is this:
- driver error; Just Don't Do That?
- worth filing an issue because it's a corner case which is worth capturing?
- surprising to anyone else?

**Comment by mpvl** [comment, marked as answer]
CUE gets the single quotes as is and interprets that as a byte sequence, rather than a selector . (Single quote strings are byte sequences in CUE.)

By default, export outputs JSON and byte sequences are output as Base64 in json. Also, CUE does not allow byte sequences to be converted to strings. Maybe it could if it is valid UTF-8, or just force conversion to UTF-8.

You can use `--out cue` to see what CUE sees before conversion. It is this expression exactly that is evaluated within the top of the config. A constant (`'foo.bar-baz'`) just evaluates to a constant. Only "references" resolve to existing values in the configuration.

**Reply by jpluscplusm** [reply]
Thank you - I appreciate the insight and detail. 

Since you're aware, I suspect it's not necessary for me to file an Issue, even if solely to document the foot-gun for other users. They can find the detail above :-)
